File: app/src/main/java/com/asmat/rolando/popularmovies/activities/MovieDetailActivity.java
package com.asmat.rolando.popularmovies.activities;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.Nullable;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.v4.app.ShareCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.asmat.rolando.popularmovies.R;
import com.asmat.rolando.popularmovies.adapters.CastLinearAdapter;
import com.asmat.rolando.popularmovies.adapters.ReviewsLinearAdapter;
import com.asmat.rolando.popularmovies.adapters.TrailersLinearAdapter;
import com.asmat.rolando.popularmovies.database.DatabaseManager;
import com.asmat.rolando.popularmovies.database.FavoriteMovie;
import com.asmat.rolando.popularmovies.database.Movie;
import com.asmat.rolando.popularmovies.database.WatchLaterMovie;
import com.asmat.rolando.popularmovies.models.AdapterOnClickHandler;
import com.asmat.rolando.popularmovies.models.Credit;
import com.asmat.rolando.popularmovies.models.Review;
import com.asmat.rolando.popularmovies.models.Video;
import com.asmat.rolando.popularmovies.viewmodels.MovieDetailsViewModel;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MovieDetailActivity extends AppCompatActivity implements AdapterOnClickHandler<Video> {

    @BindView(R.id.toolbarImage)
    ImageView backdrop;
    @BindView(R.id.iv_poster_thumbnail)
    ImageView thumbnail;
    @BindView(R.id.tv_release_date)
    TextView releaseDateLabel;
    @BindView(R.id.tv_movie_rating)
    TextView ratingLabel;
    @BindView(R.id.tv_synopsis_content)
    TextView synopsisLabel;

    // TODO raa create custom view to wrap label, rv, loader, error label, empty label
    // TODO raa use data bindings
    // Trailers
    @BindView(R.id.rv_trailers)
    RecyclerView trailers;
    @BindView(R.id.pb_trailers_loading_bar)
    ProgressBar trailersLoading;
    @BindView(R.id.tv_no_trailers)
    TextView noTrailersLabel;
    @BindView(R.id.tv_error_trailers)
    TextView trailersErrorLabel;
    private TrailersLinearAdapter trailersLinearAdapter;

    // Cast
    @BindView(R.id.cast_recycler_view)
    RecyclerView cast;
    @BindView(R.id.cast_loading)
    ProgressBar castLoading;
    @BindView(R.id.cast_empty)
    TextView noCastLabel;
    @BindView(R.id.cast_error)
    TextView castErrorLabel;
    private CastLinearAdapter castLinearAdapter;

    // Reviews
    @BindView(R.id.rv_reviews)
    RecyclerView reviews;
    @BindView(R.id.pb_reviews_loading_bar)
    ProgressBar reviewsLoading;
    @BindView(R.id.tv_no_reviews)
    TextView noReviewsLabel;
    @BindView(R.id.tv_error_reviews)
    TextView reviewsErrorLabel;
    private ReviewsLinearAdapter mReviewsLinearAdapter;

    // Star
    @BindView(R.id.star)
    ImageView star;
    // Bookmark
    @BindView(R.id.bookmark)
    ImageView bookmark;

    public final static String INTENT_EXTRA_MOVIE_ID = "MOVIE_ID";

    // ViewModel
    private MovieDetailsViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_movie_detail);
        ButterKnife.bind(this);

        Intent intent = getIntent();
        if (intent != null && intent.hasExtra(INTENT_EXTRA_MOVIE_ID)) {
            int movieID = intent.getIntExtra(INTENT_EXTRA_MOVIE_ID, 0);
            viewModel = ViewModelProviders.of(this).get(MovieDetailsViewModel.class);
            viewModel.init(movieID);
        }
        viewModel.getMovie().observe(this, new Observer<Movie>() {
            @Override
            public void onChanged(@Nullable Movie movie) {
                MovieDetailActivity.this.updateMovieDetails(movie);
            }
        });
        viewModel.getFavoriteMovie().observe(this, new Observer<FavoriteMovie>() {
            @Override
            public void onChanged(@Nullable FavoriteMovie favoriteMovie) {
                MovieDetailActivity.this.updateStarIcon(favoriteMovie);
            }
        });
        viewModel.getWatchLaterMovie().observe(this, new Observer<WatchLaterMovie>() {
            @Override
            public void onChanged(@Nullable WatchLaterMovie watchLaterMovie) {
                MovieDetailActivity.this.updateBookmarkIcon(watchLaterMovie);
            }
        });
        viewModel.getVideos().observe(this, new Observer<Video[]>() {
            @Override
            public void onChanged(@Nullable Video[] videos) {
                MovieDetailActivity.this.updateTrailers(videos);
            }
        });
        viewModel.getCredit().observe(this, new Observer<Credit>() {
            @Override
            public void onChanged(@Nullable Credit credit) {
                MovieDetailActivity.this.updateCredit(credit);
            }
        });

        viewModel.getReviews().observe(this, new Observer<Review[]>() {
            @Override
            public void onChanged(@Nullable Review[] reviews) {
                MovieDetailActivity.this.updateReviews(reviews);
            }
        });

        setup();
    }

    @Override
    public void onClick(Video trailer) {
        String url = trailer.getYouTubeURL();
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
        if(intent.resolveActivity(getPackageManager()) != null) {
            startActivity(intent);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                super.onBackPressed();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    //region User Actions

    public void onStar(View view) {
        Movie movie = MovieDetailActivity.this.viewModel.getMovie().getValue();
        if (movie == null) return;
        FavoriteMovie favoriteMovie = new FavoriteMovie(movie.getId());
        if (this.viewModel.getFavoriteMovie().getValue() == null) {
            // Add the Favorite movie
            DatabaseManager.INSTANCE.addMovie(movie);
            DatabaseManager.INSTANCE.addFavoriteMovie(favoriteMovie);
        } else {
            DatabaseManager.INSTANCE.deleteFavoriteMovie(favoriteMovie);
            // Check if it's a watch later movie, delete movie ref if it's not
            if (this.viewModel.getWatchLaterMovie().getValue() == null) {
                DatabaseManager.INSTANCE.deleteMovie(movie);
            }
        }
    }

    public void onBookmark(View view) {
        Movie movie = MovieDetailActivity.this.viewModel.getMovie().getValue();
        if (movie == null) return;
        WatchLaterMovie watchLaterMovie = new WatchLaterMovie(movie.getId());
        if (this.viewModel.getWatchLaterMovie().getValue() == null) {
            DatabaseManager.INSTANCE.addMovie(movie);
            DatabaseManager.INSTANCE.addWatchLaterMovie(watchLaterMovie);
        } else {
            DatabaseManager.INSTANCE.deleteWatchLaterMovie(watchLaterMovie);
            if (this.viewModel.getFavoriteMovie().getValue() == null) {
                DatabaseManager.INSTANCE.deleteMovie(movie);
            }
        }
    }

    public void onShare(View view) {
        String mimeType = "text/plain";
        String title = getResources().getString(R.string.share_movie);
        Movie movie = viewModel.getMovie().getValue();
        String movieTitle = movie.getTitle();
        String textToShare = getResources().getString(R.string.check_out_movie)+
                " \""+movieTitle+"\"";
        Video[] videos = trailersLinearAdapter.getData();
        if(videos != null && videos.length > 0) {
            textToShare += "\n" + videos[0].getYouTubeURL();
        }

        Intent intent = ShareCompat.IntentBuilder.from(this)
                .setChooserTitle(title)
                .setType(mimeType)
                .setText(textToShare).getIntent();
        startActivity(Intent.createChooser(intent, title));
    }

    //endregion

    /**
     * Private methods
     */

    //region Update UI

    private void updateMovieDetails(Movie movie) {
        if (movie != null) {
            Picasso.with(this).load(movie.getBackdropURL()).into(backdrop);
            Picasso.with(this).load(movie.getPosterURL()).into(thumbnail);
            CollapsingToolbarLayout collapsingToolbarLayout = findViewById(R.id.collapsingToolbar);
            collapsingToolbarLayout.setTitle(movie.getTitle());
            String formatted = movie.getReleaseDateFormatted();
            releaseDateLabel.setText(formatted);
            String rating = movie.getVoteAverage()+getString(R.string.out_of_ten);
            ratingLabel.setText(rating);
            synopsisLabel.setText(movie.getOverview());
        }
    }

    private void updateStarIcon(FavoriteMovie favoriteMovie) {
        if (favoriteMovie == null) {
            unfillStar();
        } else {
            fillStar();
        }
    }

    private void updateBookmarkIcon(WatchLaterMovie watchLaterMovie) {
        if (watchLaterMovie == null) {
            unfillBookmark();
        } else {
            fillBookmark();
        }
    }

    private void updateTrailers(Video[] videos) {
        trailersLoading.setVisibility(View.GONE);
        if (videos == null) {
            trailersErrorLabel.setVisibility(View.VISIBLE);
        } else {
            if(videos.length == 0) {
                noTrailersLabel.setVisibility(View.VISIBLE);
            } else {
                ArrayList<Video> trailers = new ArrayList<>();
                for(Video video: videos) {
                    if(video.getType().equals("Trailer")){
                        trailers.add(video);
                    }
                }
                if(trailers.size() == 0) {
                    noTrailersLabel.setVisibility(View.VISIBLE);
                } else {
                    Video[] array = trailers.toArray(new Video[0]);
                    trailersLinearAdapter.setData(array);
                    this.trailers.setVisibility(View.VISIBLE);
                }
            }
        }
    }

    private void updateCredit(Credit credit) {
        castLoading.setVisibility(View.GONE);
        if (credit == null) {
            castErrorLabel.setVisibility(View.VISIBLE);
        } else {
            if(credit.getCast().length == 0) {
                noCastLabel.setVisibility(View.VISIBLE);
            } else {
                castLinearAdapter.setData(credit.getCast());
                this.cast.setVisibility(View.VISIBLE);
            }
        }
    }

    private void updateReviews(Review[] reviews) {
        reviewsLoading.setVisibility(View.GONE);
        if(reviews == null) {
            reviewsErrorLabel.setVisibility(View.VISIBLE);
        } else {
            if(reviews.length == 0) {
                noReviewsLabel.setVisibility(View.VISIBLE);
            } else {
                this.reviews.setVisibility(View.VISIBLE);
                mReviewsLinearAdapter.setData(reviews);
            }
        }
    }

    //endregion

    //region setup

    private void setup() {
        setupToolbar();
        setupTrailersRecyclerView();
        setupReviewsRecyclerView();
        setupCastRecyclerView();
    }

    private void setupToolbar() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    private void setupTrailersRecyclerView() {
        trailers.setHasFixedSize(true);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        trailers.setLayoutManager(layoutManager);
        trailersLinearAdapter = new TrailersLinearAdapter(this);
        trailers.setAdapter(trailersLinearAdapter);
        trailers.setNestedScrollingEnabled(false);
    }

    private void setupCastRecyclerView() {
        cast.setHasFixedSize(true);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        cast.setLayoutManager(layoutManager);
        castLinearAdapter = new CastLinearAdapter();
        cast.setAdapter(castLinearAdapter);
        cast.setNestedScrollingEnabled(false);
    }

    private void setupReviewsRecyclerView() {
        reviews.setHasFixedSize(true);
        LinearLayoutManager reviewsLayoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        reviews.setLayoutManager(reviewsLayoutManager);
        reviewsLayoutManager.setSmoothScrollbarEnabled(true);
        mReviewsLinearAdapter = new ReviewsLinearAdapter();
        reviews.setAdapter(mReviewsLinearAdapter);
        reviews.setNestedScrollingEnabled(false);
    }

    //endregion

    //region Icons

    private void fillStar() {
        star.setImageResource(R.drawable.ic_star_filled);
    } // TODO raa use selectors instead

    private void unfillStar() {
        star.setImageResource(R.drawable.ic_star);
    }

    private void fillBookmark() {
        bookmark.setImageResource(R.drawable.ic_bookmark_filled);
    }

    private void unfillBookmark() {
        bookmark.setImageResource(R.drawable.ic_bookmark);
    }

    //endregion
}

File: app/src/main/java/com/asmat/rolando/popularmovies/ui/activities/MovieDetailActivity.kt
package com.asmat.rolando.popularmovies.ui.activities

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.net.Uri
import android.support.v4.app.ShareCompat
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.Toolbar
import android.view.MenuItem
import android.view.View
import com.asmat.rolando.popularmovies.MovieNightApplication

import com.asmat.rolando.popularmovies.R
import com.asmat.rolando.popularmovies.ui.adapters.linear.CastLinearAdapter
import com.asmat.rolando.popularmovies.ui.adapters.linear.ReviewsLinearAdapter
import com.asmat.rolando.popularmovies.ui.adapters.linear.TrailersLinearAdapter
import com.asmat.rolando.popularmovies.extensions.gone
import com.asmat.rolando.popularmovies.extensions.visible
import com.asmat.rolando.popularmovies.model.Movie
import com.asmat.rolando.popularmovies.repositories.MoviesRepository
import com.asmat.rolando.popularmovies.networking.the.movie.db.models.*
import com.asmat.rolando.popularmovies.viewmodels.MovieDetailsViewModel
import com.squareup.picasso.Picasso

import com.asmat.rolando.popularmovies.utilities.URLUtils
import com.asmat.rolando.popularmovies.viewmodels.ViewModelFactory
import kotlinx.android.synthetic.main.activity_movie_detail.*
import kotlinx.android.synthetic.main.movie_details_user_actions.*
import kotlinx.android.synthetic.main.primary_details.*
import javax.inject.Inject

class MovieDetailActivity : AppCompatActivity() {

    companion object {
        const val INTENT_EXTRA_MOVIE_DATA = "MOVIE_DATA"
    }

    @Inject
    lateinit var moviesRepository: MoviesRepository

    // Recycler View Adapters
    private lateinit var trailersLinearAdapter: TrailersLinearAdapter
    private lateinit var castLinearAdapter: CastLinearAdapter
    private lateinit var reviewsLinearAdapter: ReviewsLinearAdapter

    // View Model
    private lateinit var viewModel: MovieDetailsViewModel

    //region Callbacks
    private val trailerClickCallback = { trailer: VideosResponse.Video ->
        val url = URLUtils.getYoutubeURL(trailer.key)
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        intent.resolveActivity(packageManager)?.let {
            startActivity(intent)
        }
    }

    private val castClickCallback = { cast: CreditsResponse.Cast ->
        val intent = Intent(this@MovieDetailActivity, CastDetailActivity::class.java)
        intent.putExtra(CastDetailActivity.EXTRA_CAST, cast)
        intent.resolveActivity(packageManager)?.let {
//            startActivity(intent)
        }
    }
    //endregion

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (applicationContext as MovieNightApplication).component().inject(this)
        setContentView(R.layout.activity_movie_detail)

        if (intent != null && intent.hasExtra(INTENT_EXTRA_MOVIE_DATA)) {
            val movieData = intent.getParcelableExtra<Movie>(INTENT_EXTRA_MOVIE_DATA)
            viewModel = ViewModelProviders.of(this, ViewModelFactory(moviesRepository, movieData)).get(MovieDetailsViewModel::class.java)
        }
        setupObservers()
        setupUI()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                super.onBackPressed()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSupportNavigateUp(): Boolean {
        super.onBackPressed()
        return true
    }

    //region User Actions

    fun onStar(view: View) {
        viewModel.onStarTapped()
    }

    fun onBookmark(view: View) {
        viewModel.onBookmarkTapped()
    }

    fun onShare(view: View) {
        viewModel.onShareTapped()
    }

    //endregion

    //region setup

    private fun setupObservers() {
        // Movie info
        viewModel
                .backdropURL
                .observe(this, Observer { url ->
                    updateBackdrop(url)
                })

        viewModel
                .movieTitle
                .observe(this, Observer { title ->
                    updateTitle(title)
                })

        viewModel
                .releaseDate
                .observe(this, Observer { date ->
                    updateReleaseDate(date)
                })

        viewModel
                .rating
                .observe(this, Observer { rating ->
                    updateRating(rating)
                })

        viewModel
                .posterURL
                .observe(this, Observer { url ->
                    updatePoster(url)
                })

        viewModel
                .summary
                .observe(this, Observer { summary ->
                    updateSummary(summary)
                })

        // Movie icons

        viewModel
                .isFavoriteMovie
                .observe(this, Observer { isFavoriteMovie ->
                    updateStarIcon(isFavoriteMovie)
                })

        viewModel
                .isWatchLaterMovie
                .observe(this, Observer { isWatchLaterMovie ->
                    updateBookmarkIcon(isWatchLaterMovie)
                })

        viewModel
                .shareMovie
                .observe(this, Observer { movieData ->
                    shareMovie(movieData)
                })

        // Movie lists

        viewModel
                .videos
                .observe(this, Observer { videos ->
                    updateTrailers(videos)
                })

        viewModel
                .cast
                .observe(this, Observer { cast ->
                    updateCast(cast)
                })

        viewModel
                .reviews
                .observe(this, Observer { reviews ->
                    updateReviews(reviews)
                })
    }

    private fun setupUI() {
        setupToolbar()
        setupTrailersRecyclerView()
        setupReviewsRecyclerView()
        setupCastRecyclerView()
    }

    private fun setupToolbar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    private fun setupTrailersRecyclerView() {
        trailersRecyclerView.setHasFixedSize(true)
        val layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        trailersRecyclerView.layoutManager = layoutManager
        trailersLinearAdapter = TrailersLinearAdapter(trailerClickCallback)
        trailersRecyclerView.adapter = trailersLinearAdapter
        trailersRecyclerView.isNestedScrollingEnabled = false
    }

    private fun setupCastRecyclerView() {
        castRecyclerView.setHasFixedSize(true)
        val layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        castRecyclerView.layoutManager = layoutManager
        castLinearAdapter = CastLinearAdapter(castClickCallback)
        castRecyclerView.adapter = castLinearAdapter
        castRecyclerView.isNestedScrollingEnabled = false
    }

    private fun setupReviewsRecyclerView() {
        reviewsRecyclerView.setHasFixedSize(true)
        val reviewsLayoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        reviewsRecyclerView.layoutManager = reviewsLayoutManager
        reviewsLayoutManager.isSmoothScrollbarEnabled = true
        reviewsLinearAdapter = ReviewsLinearAdapter()
        reviewsRecyclerView.adapter = reviewsLinearAdapter
        reviewsRecyclerView.isNestedScrollingEnabled = false
    }

    //endregion

    //region Update UI

    private fun updateBackdrop(url: String?) {
        url?.let {
            Picasso.with(this).load(it).into(backdrop)
        }
    }

    private fun updateTitle(title: String?) {
        title?.let {
            collapsingToolbar.title = it
        }
    }

    private fun updateReleaseDate(date: String?) {
        date?.let {
            releaseDateLabel.text = it
        }
    }

    private fun updateRating(rating: String?) {
        rating?.let {
            val ratingFormatted = getString(R.string.out_of_ten, it)
            movieRatingLabel.text = ratingFormatted
        }
    }

    private fun updatePoster(url: String?) {
        url?.let {
            Picasso.with(this).load(it).into(thumbnail)
        }
    }

    private fun updateSummary(summaryStr: String?) {
        summaryStr?.let {
            summary.text = it
        }
    }

    private fun updateStarIcon(isFavoriteMovie: Boolean?) {
        isFavoriteMovie?.let {
            if (it) {
                fillStar()
            } else {
                unfillStar()
            }
        } ?: unfillStar()
    }

    private fun updateBookmarkIcon(isWatchLaterMovie: Boolean?) {
        isWatchLaterMovie?.let {
            if (it) {
                fillBookmark()
            } else {
                unfillBookmark()
            }
        } ?: unfillBookmark()
    }

    private fun updateTrailers(videos: List<VideosResponse.Video>?) { // TODO observe and update error state
        trailersLoadingBar.gone()
        when {
            videos == null -> trailersErrorStateLabel.visible()
            videos.isEmpty() -> trailersEmptyStateLabel.visible()
            else -> {
                trailersLinearAdapter.data = videos
                trailersRecyclerView.visible()
            }
        }
    }

    private fun updateCast(cast: List<CreditsResponse.Cast>?) {
        castLoadingBar.gone()
        when {
            cast == null -> castErrorStateLabel.visible()
            cast.isEmpty() -> castEmptyStateLabel.visible()
            else -> {
                castLinearAdapter.data = cast
                castRecyclerView.visible()
            }
        }
    }

    private fun updateReviews(reviews: List<ReviewsResponse.Review>?) {
        reviewsLoadingBar.gone()
        when {
            reviews == null -> reviewsErrorStateLabel.visible()
            reviews.isEmpty() -> reviewsEmptyStateLabel.visible()
            else -> {
                reviewsLinearAdapter.data = reviews
                reviewsRecyclerView.visible()
            }
        }
    }

    private fun shareMovie(movieData: Pair<String, String>?) {
        if (movieData == null) { return }
        val movieTitle = movieData.first
        val movieURL = movieData.second
        val mimeType = "text/plain"
        val title = resources.getString(R.string.share_movie)
        val textToShare = resources.getString(R.string.check_out_movie, movieTitle, movieURL)
        val intent = ShareCompat.IntentBuilder.from(this)
                .setChooserTitle(title)
                .setType(mimeType)
                .setText(textToShare).intent
        startActivity(Intent.createChooser(intent, title))
    }

    //endregion

    //region Icons

    private fun fillStar() {
        star.setImageResource(R.drawable.ic_star_filled)
    }

    private fun unfillStar() {
        star.setImageResource(R.drawable.ic_star)
    }

    private fun fillBookmark() {
        bookmark.setImageResource(R.drawable.ic_bookmark_filled)
    }

    private fun unfillBookmark() {
        bookmark.setImageResource(R.drawable.ic_bookmark)
    }

    //endregion
}